### DISCUSSION TITLE
Share a function between threads in pool

### DISCUSSION THREAD

**Discussion by RobinTail** [vitest-dev/vitest#6527] [Q&A] [answered]
Dear community,

I need your help in figuring out how to share a function among the tests running in different threads.

I have a generator* function, that returns different ports each time being called, and I multiple files/tests should use it to spin a HTTP server on that unique port.

It works fine in CI, where is a single thread, or with setting `maxThreads: 1` in vitest config, but I'd like to preserve multiple threads feature and make that function available across the threads while keeping its state.

Currently, having multiple threads, each thread gets a copy of that generator* function, so that it leads to the same ports returned and tests are failing.

I tried `globalSetup` feature and assigning that function to `global`, but that also didn't work.
And the `provide` feature can not provide functions.
Please advise.

```ts
const disposer = (function* () {
  let port = 8010;
  while (true) yield port++;
})();

export const givePort = (): number => disposer.next().value;
```

**Comment by RobinTail** [comment]
So far I found a workaround to utilize the `process.env.VITEST_POOL_ID` in generator for distributing `100` ports per thread.

```ts
const disposer = (function* () {
  let port = 8e3 + 1e2 * Number(process.env.VITEST_POOL_ID);
  while (true) yield port++;
})();

export const givePort = (): number => disposer.next().value;
```

**Reply by AriPerkkio** [reply, marked as answer]
`VITEST_POOL_ID` will be unique on each test runner thread so you can use that to derive unique numbers. Other solutions could be to create small HTTP or websocket server where you could fetch the port number. Or read it from file system - but that could run into race conditions when there's no locking system used. 

Internally Vitest uses `node:worker_thread`'s communication channel, but we don't expose it to users in any way. I've seen requests like this before, so I'm starting to think if we should provide a way for users to call main thread from test runner threads. 🤔

**Reply by AriPerkkio** [reply]
> And the provide feature can not `provide` functions.

Here's another approach to work-around this issue. As `provide` is called when creating each thread, you could increase the value in getter:

```ts
// global-setup.ts
import { GlobalSetupContext } from "vitest/node";

let port = 8000;

export function setup({ provide }: GlobalSetupContext) {
  provide("port", {
    get value() {
      return port++;
    },
  });
}

// example.test.ts
import { inject, test } from "vitest";

test("example", async () => {
  const port = inject("port").value;
});
```
